### PR TITLE
Agent config api - #1498

### PR DESCRIPTION
**Pull request by jyotisingh**
Some of this also affects the internals of pipeline config api, both of these APIs use the same mechanism to validate and save config.

### CODE REVIEW

File: common/test/unit/com/thoughtworks/go/domain/AgentConfigTest.java [hunk 1]
```diff
@@ -55,9 +57,9 @@ public void shouldValidateIpV4AndIpV6() throws Exception {
 
     @Test
     public void shouldDetectInvalidIPAddress() throws Exception {
-        shouldBeInvalid("blahinvalid", "IpAddress is invalid.");
-        shouldBeInvalid("blah.invalid", "IpAddress is invalid.");
-        shouldBeInvalid("399.0.0.1", "IpAddress is invalid.");
+        shouldBeInvalid("blahinvalid", "'blahinvalid' is an invalid IP address.");
+        shouldBeInvalid("blah.invalid", "'blah.invalid' is an invalid IP address.");
+        shouldBeInvalid("399.0.0.1", "'399.0.0.1' is an invalid IP address.");
```
Comment (ketan) [inline common/test/unit/com/thoughtworks/go/domain/AgentConfigTest.java]: Get rid of or replace all `shouldBeInvalid` and `shouldBeValid` tests with the new test?

File: config/config-api/src/com/thoughtworks/go/config/AgentConfig.java [hunk 2]
```diff
@@ -72,13 +86,13 @@ private void validateIpAddress() {
             return;
         }
         if (address.trim().isEmpty()) {
```
Comment (ketan) [inline config/config-api/src/com/thoughtworks/go/config/AgentConfig.java]: Change to `StringUtils.isBlank`?

File: config/config-api/src/com/thoughtworks/go/config/exceptions/GoConfigInvalidException.java [hunk 3]
```diff
@@ -12,23 +12,34 @@
...
+    public GoConfigInvalidException(CruiseConfig cruiseConfig, List<ConfigErrors> allErrors) {
+        this(cruiseConfig, ListUtil.join(ListUtil.map(allErrors, new ListUtil.Transformer<String>() {
+            @Override
+            public String transform(Object obj) {
+                ConfigErrors errors = (ConfigErrors) obj;
+                return ListUtil.join(errors.getAll());
+            }
+        })));
+    }
+
```
Comment (zabil) [inline config/config-api/src/com/thoughtworks/go/config/exceptions/GoConfigInvalidException.java]: Looking for usages of this class and constructor. I feel we need a better API
for e.g. to construct this exception one has to do this 
```
    private GoConfigInvalidException getGoConfigInvalidException() {
        ConfigErrors configErrors = new ConfigErrors();
        configErrors.add("command", "command cannot be empty");
        ArrayList<ConfigErrors> list = new ArrayList<ConfigErrors>();
        list.add(configErrors);
        return new GoConfigInvalidException(new BasicCruiseConfig(), list);
    }

```
(this is a test)

But other usages goes like
```
throw new GoConfigInvalidException(preprocessedConfig, new ArrayList<ConfigErrors>());
```

Also ideally we should have just wrapped/extended the ```List<ConfigErrors>``` onto something like ```AllConfigErrors``` and wrapped this method onto that

```
AllConfigErrors.errorAsString
 {Fancy loops}
```

Atleast

```
              ConfigErrors errors = (ConfigErrors) obj;
               return ListUtil.join(errors.getAll());
```
Could have been

```
              ConfigErrors errors = (ConfigErrors) obj;
               return errors.asString();
```

File: common/src/com/thoughtworks/go/validation/ConfigUpdateValidator.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+/*************************GO-LICENSE-START*********************************
+ * Copyright 2015 ThoughtWorks, Inc.
```
Comment (ketan) [inline common/src/com/thoughtworks/go/validation/ConfigUpdateValidator.java]: You mean 2016, right :)

File: config/config-server/src/com/thoughtworks/go/config/MagicalGoConfigXmlLoader.java [hunk 5]
```diff
@@ -140,7 +140,7 @@ private CruiseConfig validateCruiseConfig(CruiseConfig config) throws Exception
         try {
             List<ConfigErrors> allErrors = validate(config);
             if (!allErrors.isEmpty()) {
-                throw new GoConfigInvalidException(config, allErrors);
+                throw new GoConfigInvalidException(config, allErrors.get(0).firstError());
```
Comment (ketan) [inline config/config-server/src/com/thoughtworks/go/config/MagicalGoConfigXmlLoader.java]: Why do we throw an exception with just the first error?

File: server/src/com/thoughtworks/go/server/service/AgentConfigService.java [hunk 6]
```diff
@@ -47,45 +68,446 @@ public void register(AgentChangeListener agentChangeListener) {
...
+                    "agentConfig=" + agentConfig +
+                    '}';
+        }
 
-    public void modifyResources(AgentInstance[] agentInstances, List<TriStateSelection> selections) {
-        goConfigService.modifyResources(agentInstances, selections);
     }
 
+    @Deprecated
+    public void updateAgentResources(final String uuid, final Resources resources) {
```
Comment (ketan) [inline server/src/com/thoughtworks/go/server/service/AgentConfigService.java]: This does not seem to be used, outside of tests, can we delete this method (or move it to a test helper)?

File: server/webapp/WEB-INF/rails.new/app/controllers/api_v2/agents_controller.rb [hunk 7]
```diff
@@ -33,13 +33,16 @@ def show
...
-      agent_service.updateAgentAttributes(current_user, result, params[:uuid], params[:hostname], maybe_join(params[:resources]), maybe_join(params[:environments]), to_enabled_tristate)
+
+      agent = agent_service.updateAgentAttributes(current_user, result, params[:uuid], params[:hostname], maybe_join(params[:resources]), maybe_join(params[:environments]), to_enabled_tristate)
 
       if result.isSuccess
         load_agent
         render json_hal_v2: agent_presenter.to_hash(url_builder: self)
       else
-        render_http_operation_result(result)
+        @agent_instance = agent
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/controllers/api_v2/agents_controller.rb]: This feels a bit redundant, if the `@agent_instance` instance variable were set on line 37 above, at worst it'll be nil.

File: server/webapp/WEB-INF/rails.new/app/controllers/api_v2/agents_controller.rb [hunk 8]
```diff
@@ -84,7 +87,7 @@ def agent_presenter
     end
 
     def agent_view_model
-      com.thoughtworks.go.server.ui.AgentViewModel.new(@agent_instance, environment_config_service.environmentsFor(@agent_instance.getUuid()))
+      com.thoughtworks.go.server.ui.AgentViewModel.new(@agent_instance, environment_config_service.environmentsFor(@agent_instance.getUuid())) if @agent_instance
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/controllers/api_v2/agents_controller.rb]: Is there a need to be defensive here?

File: server/webapp/WEB-INF/rails.new/app/presenters/api_v2/agent_representer.rb [hunk 9]
```diff
@@ -83,6 +84,14 @@ def free_space
       end
     end
 
+    def errors
+      mapped_errors = agent.errors
+      if ip_errors = mapped_errors.delete('ipAddress')
+        mapped_errors['ip_address'] = ip_errors
+      end
+      mapped_errors
+    end
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/presenters/api_v2/agent_representer.rb]: @juhijariwala has a nice pattern for doing this —

```
class FooRepresenter
  error_representer('ipAddress' => 'ip_address')
end
```

The `error_representer` will take care of the translation, and defining the `errors` property that you declared above.

You may need to copy some code from `ApiV1::BaseRepresenter`

File: server/webapp/WEB-INF/rails.new/app/presenters/api_v2/base_representer.rb [hunk 10]
```diff
@@ -33,7 +33,7 @@ def property(name, options={})
     end
 
     def to_hash(*options)
-      super.deep_symbolize_keys
+      super.deep_symbolize_keys unless @represented.nil?
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/presenters/api_v2/base_representer.rb]: Why was this needed here, but not in apiv1?

File: server/webapp/localize_ja.xml [hunk 11]
```diff
@@ -780,6 +780,7 @@
     <entry key="SCM_USED_IN_PIPELINES_LINK">Show pipelines using this SCM</entry>
     <entry key="STALE_PIPELINE_CONFIG">Someone has modified the configuration for pipeline ''{0}''. Please update your copy of the config with the changes.</entry>
     <entry key="PIPELINE_CONFIG_VALIDATION_FAILED">Validations failed for pipeline ''{0}''. Please correct and resubmit.</entry>
```
Comment (ketan) [inline server/webapp/localize_ja.xml]: Duplication? Can `PIPELINE_CONFIG_VALIDATION_FAILED` be implemented via `ENTITY_CONFIG_VALIDATION_FAILED` ?

File: server/src/com/thoughtworks/go/config/EntityConfigSaveResult.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+    public T getEntityConfig() {
+        return entityConfig;
+    }
+
+    public GoConfigHolder getConfigHolder() {
+        return configHolder;
+    }
+
+    public ConfigSaveState getConfigSaveState() {
```
Comment (ketan) [inline server/src/com/thoughtworks/go/config/EntityConfigSaveResult.java]: This method seems unused.

**Comment by ketan** [comment]
@jyotisingh - done with the review. Please look at the comments and update all copyright headers with 2016.
